File: ios/Sasquatsh/ViewModels/DeckListViewModel.swift
import SwiftUI

@Observable
@MainActor
final class DeckListViewModel {
    var decks: [MtgDeck] = []
    var isLoading = false
    var error: String?
    var filterFormat: String?

    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    func loadDecks() async {
        guard let services else { return }
        isLoading = true
        error = nil
        do {
            decks = try await services.mtgDecks.getMyDecks()
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func deleteDeck(_ id: String) async {
        guard let services else { return }
        do {
            try await services.mtgDecks.deleteDeck(id: id)
            decks.removeAll { $0.id == id }
        } catch {
            self.error = error.localizedDescription
        }
    }

    var filteredDecks: [MtgDeck] {
        guard let filterFormat else { return decks }
        return decks.filter { $0.formatId == filterFormat }
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/DeckListViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.MtgDeck
import com.sasquatsh.app.services.MtgDeckService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class DeckListUiState(
    val decks: List<MtgDeck> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val filterFormat: String? = null
) {
    val filteredDecks: List<MtgDeck>
        get() = if (filterFormat != null) {
            decks.filter { it.formatId == filterFormat }
        } else {
            decks
        }
}

@HiltViewModel
class DeckListViewModel @Inject constructor(
    private val mtgDeckService: MtgDeckService
) : ViewModel() {

    private val _uiState = MutableStateFlow(DeckListUiState())
    val uiState: StateFlow<DeckListUiState> = _uiState.asStateFlow()

    fun loadDecks() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val decks = mtgDeckService.getMyDecks()
                _uiState.update { it.copy(decks = decks, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isLoading = false) }
            }
        }
    }

    fun deleteDeck(id: String) {
        viewModelScope.launch {
            try {
                mtgDeckService.deleteDeck(id)
                _uiState.update { state ->
                    state.copy(decks = state.decks.filter { it.id != id })
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun updateFilterFormat(format: String?) {
        _uiState.update { it.copy(filterFormat = format) }
    }
}
